packages/react-reconciler/src/ReactFiberCommitWork.js

```diff
@@ -1744,7 +1748,7 @@ function commitMutationEffectsOnFiber(
       recursivelyTraverseMutationEffects(root, finishedWork, lanes);
       commitReconciliationEffects(finishedWork);
 
-      if (flags & Ref) {
+      if (flags & Ref && !offscreenSubtreeWasHidden) {
```

Comment:
For the Jest unit test that @rickhanlonii added in https://github.com/facebook/react/pull/31379/commits/a2c9e7ad7b21b8d934a39f2b3f7ad54ba4c7dc0a, this is the only line that is actually necessary to fix the unit test.